### PR TITLE
e2e-test: Improve test coverage for libvirt in CAA

### PR DESCRIPTION
**Pull request by chathuryaadapa**
This PR adds comprehensive test coverage for the libvirt cloud provider, focusing on core infrastructure components and CloudInit functionality. The first commit introduces 26 new test functions covering domain management, architecture detection, capabilities retrieval, XML generation for different architectures (x86_64, aarch64, s390x), and various edge cases. The second commit adds 4 CloudInit-specific tests that validate ISO creation with empty data, large data, special characters, and vendor-data verification. Together, these changes increase test coverage from 0% to 37.6%, providing robust validation of libvirt provider functionality across different system architectures and configurations.
```
# github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/libvirt.test
ld: warning: ignoring duplicate libraries: '-lvirt-lxc', '-lvirt-qemu'
=== RUN   TestCloudInit
=== RUN   TestCloudInit/basic_cloud-init
--- PASS: TestCloudInit (0.00s)
    --- PASS: TestCloudInit/basic_cloud-init (0.00s)
=== RUN   TestInMemoryCopier
--- PASS: TestInMemoryCopier (0.00s)
=== RUN   TestCreateCloudInitVariations
=== RUN   TestCreateCloudInitVariations/empty_data
=== RUN   TestCreateCloudInitVariations/large_data
=== RUN   TestCreateCloudInitVariations/special_characters
=== RUN   TestCreateCloudInitVariations/verify_vendor_data
--- PASS: TestCreateCloudInitVariations (0.01s)
    --- PASS: TestCreateCloudInitVariations/empty_data (0.00s)
    --- PASS: TestCreateCloudInitVariations/large_data (0.00s)
    --- PASS: TestCreateCloudInitVariations/special_characters (0.00s)
    --- PASS: TestCreateCloudInitVariations/verify_vendor_data (0.00s)
=== RUN   TestLibvirtConnection
    libvirt_test.go:28: Skipping because LIBVIRT_URI is not configured
--- SKIP: TestLibvirtConnection (0.00s)
=== RUN   TestGetArchitecture
    libvirt_test.go:28: Skipping because LIBVIRT_URI is not configured
--- SKIP: TestGetArchitecture (0.00s)
=== RUN   TestCreateDomainXMLArchitectures
    libvirt_test.go:28: Skipping because LIBVIRT_URI is not configured
--- SKIP: TestCreateDomainXMLArchitectures (0.00s)
=== RUN   TestGetDeletableDiskPaths
=== RUN   TestGetDeletableDiskPaths/extract_file-backed_disks_only
=== RUN   TestGetDeletableDiskPaths/skip_disks_without_file_source
=== RUN   TestGetDeletableDiskPaths/nil_domain_returns_nil
=== RUN   TestGetDeletableDiskPaths/empty_disk_list_returns_empty_slice
=== RUN   TestGetDeletableDiskPaths/skip_disks_with_empty_file_path
--- PASS: TestGetDeletableDiskPaths (0.00s)
    --- PASS: TestGetDeletableDiskPaths/extract_file-backed_disks_only (0.00s)
    --- PASS: TestGetDeletableDiskPaths/skip_disks_without_file_source (0.00s)
    --- PASS: TestGetDeletableDiskPaths/nil_domain_returns_nil (0.00s)
    --- PASS: TestGetDeletableDiskPaths/empty_disk_list_returns_empty_slice (0.00s)
    --- PASS: TestGetDeletableDiskPaths/skip_disks_with_empty_file_path (0.00s)
=== RUN   TestGetGuestForArchType
=== RUN   TestGetGuestForArchType/find_x86_64_guest
=== RUN   TestGetGuestForArchType/find_s390x_guest
=== RUN   TestGetGuestForArchType/find_aarch64_guest
=== RUN   TestGetGuestForArchType/architecture_not_found
=== RUN   TestGetGuestForArchType/ostype_mismatch
=== RUN   TestGetGuestForArchType/empty_capabilities
--- PASS: TestGetGuestForArchType (0.00s)
    --- PASS: TestGetGuestForArchType/find_x86_64_guest (0.00s)
    --- PASS: TestGetGuestForArchType/find_s390x_guest (0.00s)
    --- PASS: TestGetGuestForArchType/find_aarch64_guest (0.00s)
    --- PASS: TestGetGuestForArchType/architecture_not_found (0.00s)
    --- PASS: TestGetGuestForArchType/ostype_mismatch (0.00s)
    --- PASS: TestGetGuestForArchType/empty_capabilities (0.00s)
=== RUN   TestLookupMachine
=== RUN   TestLookupMachine/find_machine_with_canonical_name
=== RUN   TestLookupMachine/find_machine_without_canonical_name
=== RUN   TestLookupMachine/machine_not_found_returns_empty_string
=== RUN   TestLookupMachine/s390x_machine_with_canonical
=== RUN   TestLookupMachine/empty_machine_list
--- PASS: TestLookupMachine (0.00s)
    --- PASS: TestLookupMachine/find_machine_with_canonical_name (0.00s)
    --- PASS: TestLookupMachine/find_machine_without_canonical_name (0.00s)
    --- PASS: TestLookupMachine/machine_not_found_returns_empty_string (0.00s)
    --- PASS: TestLookupMachine/s390x_machine_with_canonical (0.00s)
    --- PASS: TestLookupMachine/empty_machine_list (0.00s)
=== RUN   TestGetCanonicalMachineName
=== RUN   TestGetCanonicalMachineName/find_canonical_machine_in_arch_machines
=== RUN   TestGetCanonicalMachineName/find_canonical_machine_in_domain_machines
=== RUN   TestGetCanonicalMachineName/machine_not_found_returns_error
=== RUN   TestGetCanonicalMachineName/architecture_not_found_returns_error
--- PASS: TestGetCanonicalMachineName (0.00s)
    --- PASS: TestGetCanonicalMachineName/find_canonical_machine_in_arch_machines (0.00s)
    --- PASS: TestGetCanonicalMachineName/find_canonical_machine_in_domain_machines (0.00s)
    --- PASS: TestGetCanonicalMachineName/machine_not_found_returns_error (0.00s)
    --- PASS: TestGetCanonicalMachineName/architecture_not_found_returns_error (0.00s)
=== RUN   TestCreateCloudInitISO
=== RUN   TestCreateCloudInitISO/valid_cloud-init_data
2026/05/18 10:29:31 [adaptor/cloud/libvirt] Create cloudInit iso
=== RUN   TestCreateCloudInitISO/empty_user_data
2026/05/18 10:29:31 [adaptor/cloud/libvirt] Create cloudInit iso
=== RUN   TestCreateCloudInitISO/large_user_data
2026/05/18 10:29:31 [adaptor/cloud/libvirt] Create cloudInit iso
--- PASS: TestCreateCloudInitISO (0.01s)
    --- PASS: TestCreateCloudInitISO/valid_cloud-init_data (0.00s)
    --- PASS: TestCreateCloudInitISO/empty_user_data (0.00s)
    --- PASS: TestCreateCloudInitISO/large_user_data (0.00s)
=== RUN   TestConfigVerifier
=== RUN   TestConfigVerifier/empty_URI_fails
=== RUN   TestConfigVerifier/empty_pool_name_fails
=== RUN   TestConfigVerifier/empty_network_name_fails
=== RUN   TestConfigVerifier/empty_volume_name_fails
=== RUN   TestConfigVerifier/zero_CPU_fails
=== RUN   TestConfigVerifier/zero_memory_fails
=== RUN   TestConfigVerifier/valid_config_passes
--- PASS: TestConfigVerifier (0.00s)
    --- PASS: TestConfigVerifier/empty_URI_fails (0.00s)
    --- PASS: TestConfigVerifier/empty_pool_name_fails (0.00s)
    --- PASS: TestConfigVerifier/empty_network_name_fails (0.00s)
    --- PASS: TestConfigVerifier/empty_volume_name_fails (0.00s)
    --- PASS: TestConfigVerifier/zero_CPU_fails (0.00s)
    --- PASS: TestConfigVerifier/zero_memory_fails (0.00s)
    --- PASS: TestConfigVerifier/valid_config_passes (0.00s)
=== RUN   TestGetLaunchSecurityTypeInvalidURI
--- PASS: TestGetLaunchSecurityTypeInvalidURI (0.04s)
=== RUN   TestCreateDomainXMLArchitecturesWithMocks
=== RUN   TestCreateDomainXMLArchitecturesWithMocks/s390x_architecture
=== RUN   TestCreateDomainXMLArchitecturesWithMocks/aarch64_architecture
--- PASS: TestCreateDomainXMLArchitecturesWithMocks (0.00s)
    --- PASS: TestCreateDomainXMLArchitecturesWithMocks/s390x_architecture (0.00s)
    --- PASS: TestCreateDomainXMLArchitecturesWithMocks/aarch64_architecture (0.00s)
=== RUN   TestCreateDomainXMLx86_64
=== RUN   TestCreateDomainXMLx86_64/basic_x86_64_domain
=== RUN   TestCreateDomainXMLx86_64/x86_64_with_firmware
=== RUN   TestCreateDomainXMLx86_64/unsupported_security_type
--- PASS: TestCreateDomainXMLx86_64 (0.00s)
    --- PASS: TestCreateDomainXMLx86_64/basic_x86_64_domain (0.00s)
    --- PASS: TestCreateDomainXMLx86_64/x86_64_with_firmware (0.00s)
    --- PASS: TestCreateDomainXMLx86_64/unsupported_security_type (0.00s)
=== RUN   TestCreateDomainXML
=== RUN   TestCreateDomainXML/s390x_architecture
=== RUN   TestCreateDomainXML/aarch64_architecture
=== RUN   TestCreateDomainXML/x86_64_architecture_(default)
--- PASS: TestCreateDomainXML (0.00s)
    --- PASS: TestCreateDomainXML/s390x_architecture (0.00s)
    --- PASS: TestCreateDomainXML/aarch64_architecture (0.00s)
    --- PASS: TestCreateDomainXML/x86_64_architecture_(default) (0.00s)
=== RUN   TestVerifyDomainXMLIOMMU
=== RUN   TestVerifyDomainXMLIOMMU/s390x_with_proper_IOMMU_on_disks_and_interfaces
=== RUN   TestVerifyDomainXMLIOMMU/s390x_missing_IOMMU_on_disk
=== RUN   TestVerifyDomainXMLIOMMU/aarch64_missing_IOMMU_on_interface
=== RUN   TestVerifyDomainXMLIOMMU/x86_64_does_not_require_IOMMU
--- PASS: TestVerifyDomainXMLIOMMU (0.00s)
    --- PASS: TestVerifyDomainXMLIOMMU/s390x_with_proper_IOMMU_on_disks_and_interfaces (0.00s)
    --- PASS: TestVerifyDomainXMLIOMMU/s390x_missing_IOMMU_on_disk (0.00s)
    --- PASS: TestVerifyDomainXMLIOMMU/aarch64_missing_IOMMU_on_interface (0.00s)
    --- PASS: TestVerifyDomainXMLIOMMU/x86_64_does_not_require_IOMMU (0.00s)
PASS
coverage: 19.3% of statements
ok      github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers/libvirt        (cached)        coverage: 19.3% of statements
```

### CODE REVIEW

**Comment by stevenhorsman** [comment]
I've moved this to draft as the new unit tests are failing. Please rebase and fix the tests before making it ready to require. Thanks

**Comment by stevenhorsman** [comment]
Can you fix the golangci-lint error please:
```
Error: src/cloud-providers/src/cloud-providers/libvirt/libvirt_test.go:941:4: QF1003: could use tagged switch on tt.arch (staticcheck)
			if tt.arch == "s390x" {
```

File: src/cloud-providers/libvirt/cloudinit_test.go [hunk 1]
```diff
@@ -87,3 +87,97 @@ func TestInMemoryCopier(t *testing.T) {
...
 }
+
+func TestCreateCloudInitWithEmptyData(t *testing.T) {
+	userDataContent := []byte("")
+	metaDataContent := []byte("")
+
+	isoData, err := createCloudInit(userDataContent, metaDataContent)
+	require.NoError(t, err)
+	assert.NotNil(t, isoData)
+	assert.Greater(t, len(isoData), 0)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/cloudinit_test.go]: Maybe we could combine this to `assert.NotEmpty(t, isoData)` here and elsewhere?

File: src/cloud-providers/libvirt/cloudinit_test.go [hunk 2]
```diff
@@ -87,3 +87,97 @@ func TestInMemoryCopier(t *testing.T) {
...
+	defer isoFile.Close()
+
+	isoImg, err := iso9660.OpenImage(isoFile)
+	require.NoError(t, err)
+
+	rootFile, err := isoImg.RootDir()
+	require.NoError(t, err)
+
+	children, err := rootFile.GetChildren()
+	require.NoError(t, err)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/cloudinit_test.go]: This pattern seems to be repeated. What about a helper method like:
```
func verifyISOContents(t *testing.T, isoData []byte) map[string][]byte {
    t.Helper()
    
    file, err := os.CreateTemp("", "CloudInit-*.iso")
    require.NoError(t, err)
    defer os.Remove(file.Name())
    
    err = os.WriteFile(file.Name(), isoData, os.ModePerm)
    require.NoError(t, err)
    
    isoFile, err := os.Open(file.Name())
    require.NoError(t, err)
    defer isoFile.Close()
    
    isoImg, err := iso9660.OpenImage(isoFile)
    require.NoError(t, err)
    
    rootFile, err := isoImg.RootDir()
    require.NoError(t, err)
    
    children, err := rootFile.GetChildren()
    require.NoError(t, err)
    
    files := make(map[string][]byte)
    for _, child := range children {
        data, err := io.ReadAll(child.Reader())
        require.NoError(t, err)
        files[child.Name()] = data
    }
    
    return files
}
```
(from Bob) to reduce this?

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 3]
```diff
@@ -303,3 +666,432 @@ func TestConfigVerifier(t *testing.T) {
...
+		Guests: []libvirtxml.CapsGuest{
+			{
+				OSType: "hvm",
+				Arch: libvirtxml.CapsGuestArch{
+					Name:     "x86_64",
+					Emulator: "/usr/bin/qemu-system-x86_64",
+				},
+			},
+		},
+	}
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go]: Could we extract out something like:
```
func createMockCaps(arch, emulator string, machines ...libvirtxml.CapsGuestMachine) *libvirtxml.Caps {
    return &libvirtxml.Caps{
        Guests: []libvirtxml.CapsGuest{
            {
                OSType: "hvm",
                Arch: libvirtxml.CapsGuestArch{
                    Name:     arch,
                    Emulator: emulator,
                    Machines: machines,
                },
            },
        },
    }
}
```
as this pattern is duplicated a fair few times.

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 4]
```diff
@@ -303,3 +666,432 @@ func TestConfigVerifier(t *testing.T) {
...
+		{
+			name: "basic x86_64 domain",
+			cfg: &domainConfig{
+				name:        "test-x86-vm",
+				cpu:         2,
+				mem:         2048,
+				networkName: "default",
+				bootDisk:    "/var/lib/libvirt/images/root.qcow2",
+				cidataDisk:  "/var/lib/libvirt/images/cidata.iso",
+			},
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go]: This pattern is repeated as well, so:
```
func createTestDomainConfig(name string, cpu, mem uint, networkName, cidataDisk string) *domainConfig {
    return &domainConfig{
        name:        name,
        cpu:         cpu,
        mem:         mem,
        networkName: networkName,
        bootDisk:    "/var/lib/libvirt/images/root.qcow2",
        cidataDisk:  cidataDisk,
    }
}
```
might help?

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 5]
```diff
@@ -9,6 +9,8 @@ import (
 
 	provider "github.com/confidential-containers/cloud-api-adaptor/src/cloud-providers"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	libvirt "libvirt.org/go/libvirt"
 	libvirtxml "libvirt.org/go/libvirtxml"
 )
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go:16]: Could we define some common hardcoded values as constants here:
```
const (
    testBootDisk     = "/var/lib/libvirt/images/root.qcow2"
    testCiDataISO    = "/var/lib/libvirt/images/cidata.iso"
    testCloudInitISO = "/var/lib/libvirt/images/cloudinit.iso"
    testNetworkName  = "default"
)
```

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 6]
```diff
@@ -303,3 +666,432 @@ func TestConfigVerifier(t *testing.T) {
...
+			expectError: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := verifyDomainXML(tt.domain)
+			if tt.expectError {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.errorMsg)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go]: Please replace this with assert.ErrorContains

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 7]
```diff
@@ -194,16 +207,366 @@ func TestGetDeletableDiskPaths(t *testing.T) {
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			guest, err := getGuestForArchType(tt.caps, tt.arch, tt.ostype)
+			if tt.expectError {
+				assert.Error(t, err)
+				assert.Nil(t, guest)
+				assert.Contains(t, err.Error(), "could not find any guests")
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go]: And here

File: src/cloud-providers/libvirt/libvirt_test.go [hunk 8]
```diff
@@ -89,67 +91,61 @@ func TestCreateDomainXMLs390x(t *testing.T) {
...
+			}
 
-	domCfg, err := createDomainXML(client, &domainCfg, &vm)
-	if err != nil {
-		t.Error(err)
-	}
+			domCfg, err := createDomainXML(client, &domainCfg, &vm)
+			if err != nil {
+				t.Error(err)
+			}
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/libvirt_test.go]: Please switch this pattern out for `assert.NoError(t, err)` (or require if in setup)

File: src/cloud-providers/libvirt/libvirt_test.go (hunk 3 above)
Comment (stevenhorsman) [reply src/cloud-providers/libvirt/libvirt_test.go]: You don't seem to have replaced all of these?

File: src/cloud-providers/libvirt/libvirt_test.go (hunk 5 above)
Comment (stevenhorsman) [reply src/cloud-providers/libvirt/libvirt_test.go:16]: You've not updated these?

Comment (chathuryaadapa) [reply src/cloud-providers/libvirt/libvirt_test.go:16]: could you please check [this](https://github.com/confidential-containers/cloud-api-adaptor/pull/3051/changes#diff-ac97db25e088a7daf55678f77afaea4981deda253c63af49188fe675f82bfe59R17-R22) I did update it actually and made use of those in the code

Comment (stevenhorsman) [reply src/cloud-providers/libvirt/libvirt_test.go:16]: I was referring to the other places where it hasn't been used e.g.
- https://github.com/chathuryaadapa/cloud-api-adaptor/blob/cdd7615743e257a8eb10c63087b326a7870a4f6a/src/cloud-providers/libvirt/libvirt_test.go#L139
- https://github.com/chathuryaadapa/cloud-api-adaptor/blob/cdd7615743e257a8eb10c63087b326a7870a4f6a/src/cloud-providers/libvirt/libvirt_test.go#L171
- https://github.com/chathuryaadapa/cloud-api-adaptor/blob/cdd7615743e257a8eb10c63087b326a7870a4f6a/src/cloud-providers/libvirt/libvirt_test.go#L194
- https://github.com/chathuryaadapa/cloud-api-adaptor/blob/cdd7615743e257a8eb10c63087b326a7870a4f6a/src/cloud-providers/libvirt/libvirt_test.go#L205
- https://github.com/chathuryaadapa/cloud-api-adaptor/blob/cdd7615743e257a8eb10c63087b326a7870a4f6a/src/cloud-providers/libvirt/libvirt_test.go#L708

**Review by stevenhorsman — APPROVED** [review]
LGTM. Thanks

**Review by mkulke — APPROVED** [review]
